### PR TITLE
also resolves #9, channelized interfaces

### LINKED ISSUE #9: Add a way to search by prefix
It is often useful to look up the set of nodes with a common prefix.

### PR DESCRIPTION
**Pull request by fredim**
on top of https://github.com/kellydunn/go-art/pull/10 adds a more idiomatic interface to the iterators, using channels as opposed to callbacks.  kept the original api for backwards compatibility

### CODE REVIEW

File: art_tree.go [hunk 1]
```diff
@@ -17,40 +20,90 @@ func NewArtTree() *ArtTree {
...
+}
+
 // Returns the node that contains the passed in key, or nil if not found.
 func (t *ArtTree) Search(key []byte) interface{} {
 	key = ensureNullTerminatedKey(key)
-	return t.searchHelper(t.root, key, 0)
+	foundNode := t.searchHelper(t.root, key, 0)
+	if foundNode != nil {
+		// i think with the null terminated key, the return is always a leaf, or nil?
+		// if foundNode.IsLeaf() {
```
Comment (kellydunn) [inline art_tree.go]: Correct, `searchHelper` will only ever return the resultant node or `nil` if it does not find a match.  LeafNodes are the only ones that can match a search term.

Would you mind removing this comment as a part of your PR?  Thanks!

File: art_tree.go [hunk 2]
```diff
@@ -3,10 +3,13 @@ package art
 
 import (
 	"bytes"
+	"fmt"
 	_ "math"
 	_ "os"
 )
 
+var _ = fmt.Println
```
Comment (kellydunn) [inline art_tree.go]: This doesn't appear to be used anywhere; is this variable and the import of `fmt` necessary?  If not, I'd like to remove it :)

File: art_tree.go [hunk 3]
```diff
@@ -17,40 +20,90 @@ func NewArtTree() *ArtTree {
...
+}
+
+// Finds the starting node for a prefix search and returns an array of all the objects under it
+func (t *ArtTree) PrefixSearch(key []byte) []interface{} {
+	ret := make([]interface{}, 0)
+	for r := range t.PrefixSearchChan(key) {
+		ret = append(ret, r.Value)
+	}
+	return ret
+}
```
Comment (kellydunn) [inline art_tree.go:51]: This is an interesting approach.  It's not formally specified in the research paper I used to base this data structure on, but I think it could be a neat way of being able to get a subtree of values, should others want a specific prefix.

Curious, is there a reason why you chose to append all of the leaf nodes together instead of just returning a subtree?

File: art_tree.go [hunk 4]
```diff
@@ -17,40 +20,90 @@ func NewArtTree() *ArtTree {
...
+			for n := range t.EachChanFrom(start) {
+				if n.IsLeaf() {
+					outChan <- Result{string(n.key), n.value}
+				}
+			}
+		}
+		close(outChan)
+	}()
+	return outChan
+}
```
Comment (kellydunn) [inline art_tree.go:42]: Could you elaborate a bit more on the desire to have a channel based approach here?  It seems to me that this code is returning a channel while it asynchronously processes the art tree and appends values to the outChan.  

Could the returned channel ever have lock contention?  What if I wanted to use it immediately after calling `EachChanResultFrom`?

Would this method be easier to use as a client if a `[]Result` was returned instead?

File: art_tree.go [hunk 5]
```diff
@@ -17,40 +20,90 @@ func NewArtTree() *ArtTree {
 	return &ArtTree{root: nil, size: 0}
 }
 
+type Result struct {
+	Key   string
+	Value interface{}
+}
```
Comment (kellydunn) [inline art_tree.go:23]: Any reason not to just use `LeafNode` here?

**Comment by kellydunn** [comment]
Heya @fredim!  Thanks for the PR, I really appreciate it :) 

And my apologies on the slow response time!  Things have been a bit crazy for me lately.

I've left a few comments on your PR if you don't mind.  Would love to hear back from you!

File: art_tree.go (hunk 5 above)
Comment (fredim) [inline art_tree.go:23]: perhaps, but the key is not exposed.  the only other solution (this) is to make a special result struct so that the receiver can see what the key was.

File: art_tree.go (hunk 4 above)
Comment (fredim) [inline art_tree.go:42]: this converts from an object factory model to a transmit-receive.  the channel is open until it's closed, while the receiver waits for data and the returned channel is immediately usable.  even better, processing the results onto the channel are moved into the background.

this method is easier, especially if you're not going to build an array.  i'll point to the test case below.

if you were going to collect the results into an array, then this would be overkill.  which is why the single Search returns in the actual result.  but in a prefix search, we're expecting a series of matches, and then we'll pick over them or reassemble downstream.

File: art_tree.go (hunk 3 above)
Comment (fredim) [inline art_tree.go:51]: in this case, i left this in as a convenience.  you'd probably be better off receiving the results on a channel.  the subtree is an interesting idea, but i like the cleaner separation of the library as an api.  with that, there are some private members that are unaccessible.  but then, the calling code is much nicer.

File: art_tree_test.go [hunk 6]
```diff
@@ -729,8 +729,111 @@ func TestInsertWithSameByteSliceAddress(t *testing.T) {
...
+	rr = tree.PrefixSearch([]byte("ax"))
+	if rr == nil {
+		t.Error("empty results empty arrays")
+	} else if len(rr) > 0 {
+		t.Error("shouldn't be found", rr)
+	}
+
+	rs := ""
+	for res := range tree.PrefixSearchChan([]byte("abc")) {
+		rs += res.Value.(string) + ","
```
Comment (fredim) [inline art_tree_test.go:771]: all of the test cases process the prefix searches as a series of results.  but all of the other ones (i only modified this one to use the channel), you can see we made an array of results somewhere, but only used it to scan.  this channel case is more interesting, because we are receiving the results and processing them inline.

**Comment by fredim** [comment]
thanks @kellydunn ! i appreciate the look.  i made the mods requested, and added some comments.

File: art_tree.go (hunk 5 above)
Comment (kellydunn) [inline art_tree.go:23]: I think there was a PR a few months ago that was never completed that exposed the Key in the Leaf Node.  (https://github.com/kellydunn/go-art/pull/8#issuecomment-151049436)  I think that waiting for that to merge and get finished up might be better before introducing a new exported struct.  What do you think?  Would that be sufficient for your purposes?

File: art_tree.go (hunk 4 above)
Comment (kellydunn) [inline art_tree.go:42]: Hey @fredim, thanks for clarifying a bit further!  I have some more responses below:

> if you were going to collect the results into an array, then this would be overkill. 

I'd like to know a little bit more why you think this might be overkill.  I think we could simplify this method to the following if we just used an array as a return value:

``` go
func (t *ArtTree) EachChanResultFrom(start *ArtNode) []Result {
  children := t.Children()
  res := []Result{}
  for _, e := range children {
    if e.IsLeaf() {
      res = append(res, e)
    }
  }

  return res
}
```

> this method is easier, especially if you're not going to build an array.

I would argue that this type of approach puts a bit more burden on the client of the library.  For example, when using graphs, trees, or other data structures, you typically want immediate feedback, like so:

``` go
results := t.PrefixSearch("abc")
fmt.Printf(len(results))
// Perform Binary Search on results / reduce them / filter them as needed
// Display to the user how many matches the current prefix has
```

Asynchronous processing on the other hand, pushes the responsibility on the client to tally up results, wait until the channel returns a zero value so it knows that it's closed, and effectively write some boiler plate code that requires concurrent mechanisms to manage all of this.

For example, if I was to use this as a client of `go-art`:

``` go
outChan := t.EachResultFrom(t.Root)

for {
  res := <- outChan
  if res == Result{} {
    break
  }

  // do my operation with res
}

// At this point I know have all my results
```

Overall, I'm not 100% convinced that this is the direction the library should go in, mainly because I want to keep the interface as simple as possible for the consumer.  I don't see a clear reason why we need asynchronous processing just yet, but I appreciate your take on it!  

# 

tl;dr: Using channels is a neat concept, but I think it could lead to code bloat on the client's side, and I'd rather just return an array of results for the time being.  It's less concurrent management of resources, and it provides immediate feedback.  Given the speed optimizations of ART trees in general, I'm not sure we absolutely need concurrent processing of results at this time.

File: art_tree.go (hunk 5 above)
Comment (nickpalmer) [inline art_tree.go:23]: Sorry for the slow progress on that. I got pulled into other work and have only just gotten down far enough to context switch back. Prefix search was going to be my NEXT PR, as that is what I really need, so I will take a look at this one now.
